test/api/test-ot-color.c

```diff
@@ -0,0 +1,318 @@
+  12 | 0000 0002                      # colorRecordIndex=[0, 2]
+  16 | 000000ff ffcc66ff              # colorRecord #0, #1 (BGRA)
+  24 | 000000ff 000080ff              # colorRecord #2, #3 (BGRA)
+ */
+static hb_face_t *cpal_v0 = NULL;
+
+/* Test font with the following CPAL v1 table, as TTX and manual disassembly:
+
+  <CPAL>
+    <version value="1"/>
+    <numPaletteEntries value="2"/>
+    <palette index="0" label="257" type="2">
+      <color index="0" value="#000000FF"/>
+      <color index="1" value="#66CCFFFF"/>
+    </palette>
+    <palette index="1" label="65535" type="1">
+      <color index="0" value="#000000FF"/>
+      <color index="1" value="#FFCC66FF"/>
+    </palette>
+    <palette index="2" label="258" type="0">
+      <color index="0" value="#000000FF"/>
+      <color index="1" value="#800000FF"/>
+    </palette>
+    <paletteEntryLabels>
+      <label index="0" value="65535"/>
+      <label index="1" value="256"/>
+    </paletteEntryLabels>
+  </CPAL>
+
+   0 | 0001                           # version=1
+   2 | 0002                           # numPaletteEntries=2
+   4 | 0003                           # numPalettes=3
+   6 | 0006                           # numColorRecords=6
+   8 | 0000001e                       # offsetToFirstColorRecord=30
+  12 | 0000 0002 0004                 # colorRecordIndex=[0, 2, 4]
+  18 | 00000036                       # offsetToPaletteTypeArray=54
+  22 | 00000042                       # offsetToPaletteLabelArray=66
+  26 | 00000048                       # offsetToPaletteEntryLabelArray=72
+  30 | 000000ff ffcc66ff 000000ff     # colorRecord #0, #1, #2 (BGRA)
+  42 | 66ccffff 000000ff 000080ff     # colorRecord #3, #4, #5 (BGRA)
+  54 | 00000002 00000001 00000000     # paletteFlags=[2, 1, 0]
+  66 | 0101 ffff 0102                 # paletteName=[257, 0xffff, 258]
+  72 | ffff 0100                      # paletteEntryLabel=[0xffff, 256]
+*/
+static hb_face_t *cpal_v1 = NULL;
+
+
+#define assert_color_rgba(colors, i, r, g, b, a) G_STMT_START {	\
+  const hb_ot_color_t *_colors = (colors); \
+  const size_t _i = (i); \
+  const uint8_t red = (r), green = (g), blue = (b), alpha = (a); \
+  if (_colors[_i].red != r) { \
+    g_assertion_message_cmpnum (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+				"colors[" #i "].red", _colors[_i].red, "==", red, 'x'); \
+  } \
+  if (colors[i].green != green) { \
+    g_assertion_message_cmpnum (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+				"colors[" #i "].green", colors[i].green, "==", green, 'x'); \
+  } \
+  if (colors[i].blue != blue) { \
+    g_assertion_message_cmpnum (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+				"colors[" #i "].blue", colors[i].blue, "==", blue, 'x'); \
+  } \
+  if (colors[i].alpha != alpha) { \
+    g_assertion_message_cmpnum (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
+				"colors[" #i "].alpha", colors[i].alpha, "==", alpha, 'x'); \
+  } \
+} G_STMT_END
```

Comment:
This macro currently has an effort at safety but doesn’t go all the way. To make it safer, the red condition should use `red` instead of `r`; the green, blue, and alpha conditions should use `_colors[_i]` instead of `colors[i]`; and the green, blue, and alpha assertions should also use `_colors[_i]` instead of `colors[i]`.